app/ante/cosmos/fees.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2022 Evmos Foundation
+// This file is part of the Evmos Network packages.
+//
+// Evmos is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Evmos packages are distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Evmos packages. If not, see https://github.com/evmos/evmos/blob/main/LICENSE
+
+package cosmos
+
+import (
+	"fmt"
+	"math"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	anteutils "github.com/evmos/evmos/v11/app/ante/utils"
+)
+
+// DeductFeeDecorator deducts fees from the first signer of the tx.
+// If the first signer does not have the funds to pay for the fees,
+// and does not have enough unclaimed staking rewards, then return
+// with InsufficientFunds error.
+// The next AnteHandler is called if fees are successfully deducted.
+//
+// CONTRACT: Tx must implement FeeTx interface to use DeductFeeDecorator
+type DeductFeeDecorator struct {
+	accountKeeper      authante.AccountKeeper
+	bankKeeper         BankKeeper
+	distributionKeeper anteutils.DistributionKeeper
+	feegrantKeeper     authante.FeegrantKeeper
+	stakingKeeper      anteutils.StakingKeeper
+	txFeeChecker       anteutils.TxFeeChecker
+}
+
+// NewDeductFeeDecorator returns a new DeductFeeDecorator.
+func NewDeductFeeDecorator(
+	ak authante.AccountKeeper,
+	bk BankKeeper,
+	dk anteutils.DistributionKeeper,
+	fk authante.FeegrantKeeper,
+	sk anteutils.StakingKeeper,
+	tfc anteutils.TxFeeChecker,
+) DeductFeeDecorator {
+	if tfc == nil {
+		tfc = checkTxFeeWithValidatorMinGasPrices
+	}
+
+	return DeductFeeDecorator{
+		accountKeeper:      ak,
+		bankKeeper:         bk,
+		distributionKeeper: dk,
+		feegrantKeeper:     fk,
+		stakingKeeper:      sk,
+		txFeeChecker:       tfc,
+	}
+}
+
+// AnteHandle ensures that the transaction contains valid fee requirements and tries to deduct those
+// from the account balance or unclaimed staking rewards, which the transaction sender might have.
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, errorsmod.Wrap(errortypes.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	if !simulate && ctx.BlockHeight() > 0 && feeTx.GetGas() == 0 {
+		return ctx, errorsmod.Wrap(errortypes.ErrInvalidGasLimit, "must provide positive gas")
```

Comment:
Is this really checking for positive gas? - as we have seen in other places, it might be worthchecking that the amount of gas is really not negative.